runtime/unix.c

```diff
@@ -427,6 +429,28 @@ char *caml_secure_getenv (char const *var)
 #endif
 }
 
+int64_t caml_time_counter(void)
+{
+#if defined(_POSIX_TIMERS)                      \
+  && defined(_POSIX_MONOTONIC_CLOCK)            \
+  && _POSIX_MONOTONIC_CLOCK != (-1)
+
+  struct timespec t;
+  clock_gettime(CLOCK_MONOTONIC, &t);
+  return
+    (int64_t)t.tv_sec  * (int64_t)1000000000 +
+    (int64_t)t.tv_nsec;
+#elif defined(HAS_GETTIMEOFDAY)
+  struct timeval t;
+  gettimeofday(&t, 0);
+  return
+    (int64_t)t.tv_sec  * (int64_t)1000000000 +
+    (int64_t)t.tv_usec * (int64_t)1000;
+#else
+# error "No timesource available"
```

Comment:
On Windows, I think you can use the functions [QueryPerformanceCounter](https://docs.microsoft.com/en-us/windows/win32/api/profileapi/nf-profileapi-queryperformancecounter) + [QueryPerformanceFrequency](https://docs.microsoft.com/en-us/windows/win32/api/profileapi/nf-profileapi-queryperformancefrequency).

Reply:
Also please don't fallback on `gettimeofday` it can go back in time which can only mislead. On macOS use `mach_absolute_time`, `clock_gettime` is not supported. Here's [code](https://github.com/b0-system/b0/blob/b00/src-std/b0_stubs_time.c) for this that supports all major platforms, IIRC the windows path has been tested at some point.

Reply:
Thank you very much, I will as well test the Windows codepath.
@kayceesrk this function was imported from OCaml multicore, should I backport it later?

Reply:
I reused the code for OSX and Windows, and it seems to work fine after testing it on both platforms.
One thing I'm still unsure about is how to handle failure and how to handle other platforms:
For now it returns 0 on "unsupported" and failure (initialization of state in win32 and MacOS).
I'm thinking that the right way to do it would be to have every tracing code behind macros and be explicit we request a proper monotonic clock for it to work.

Reply:
> For now it returns 0 on "unsupported"

For that one you could at least emit a [compilation warning](https://github.com/b0-system/b0/blob/b00/src-std/b0_stubs_time.c#L79).

> I'm thinking that the right way to do it would be to have every tracing code behind macros and be explicit we request a proper monotonic clock for it to work.

Not sure I understand what you mean by that and how it solves the problem. It seems to me that returning `0` on failure is ok. No program correctness depends on that and it should make it quite clear when perusing the resulting data set you get that something went wrong.